File: packages/reference/src/state.ts
/**
 * Fold an ordered event stream into the current state of an arc: the cheap
 * "what is true now?" reduction that design-rationale.md promises the reference
 * implementation provides. Append-only events compose with any transport; this
 * reconstructs current state for consumers that want a snapshot rather than a
 * timeline. Pure, zero dependencies.
 */

import type { ArcStatusEvent, ArcStatusPhase } from "./types.js";

export interface ArcMilestone {
  title: string;
  sent_at: string;
  step?: number;
  total?: number;
}

export interface ArcState {
  arc_id: string;
  /** Objective, taken from the `started` event (falls back to the first event). */
  title: string;
  /** Phase of the most recent event. */
  phase: ArcStatusPhase;
  /** Derived lifecycle status. `blocked` clears once a later non-blocked event arrives. */
  status: "active" | "blocked" | "done";
  /** Latest `step`/`total`/`eta_minutes` seen on any event, if present. */
  step?: number;
  total?: number;
  eta_minutes?: number;
  startedAt: string;
  lastEventAt: string;
  eventCount: number;
  milestones: ArcMilestone[];
  /** The current blocker if the most recent event is `blocked`, else null. */
  blocked: { title: string; body?: string; sent_at: string } | null;
}

/**
 * Reduce an ordered sequence of events for a single arc into current state.
 * Returns null for an empty sequence. Does not validate the sequence; pass a
 * sequence you trust (or run `validateSequence` first).
 */
export function reduceArc(events: ArcStatusEvent[]): ArcState | null {
  if (events.length === 0) return null;

  const first = events[0]!;
  const last = events[events.length - 1]!;
  const started = events.find((e) => e.phase === "started");

  const milestones: ArcMilestone[] = [];
  let step: number | undefined;
  let total: number | undefined;
  let eta_minutes: number | undefined;

  for (const e of events) {
    if (e.phase === "milestone") {
      milestones.push({
        title: e.title,
        sent_at: e.sent_at,
        step: e.step,
        total: e.total,
      });
    }
    if (typeof e.step === "number") step = e.step;
    if (typeof e.total === "number") total = e.total;
    if (typeof e.eta_minutes === "number") eta_minutes = e.eta_minutes;
  }

  const status: ArcState["status"] =
    last.phase === "done"
      ? "done"
      : last.phase === "blocked"
        ? "blocked"
        : "active";

  const blocked =
    last.phase === "blocked"
      ? { title: last.title, body: last.body, sent_at: last.sent_at }
      : null;

  return {
    arc_id: first.arc_id,
    title: started?.title ?? first.title,
    phase: last.phase,
    status,
    step,
    total,
    eta_minutes,
    startedAt: started?.sent_at ?? first.sent_at,
    lastEventAt: last.sent_at,
    eventCount: events.length,
    milestones,
    blocked,
  };
}

File: reference/python/src/agent_arc_status/state.py
"""Fold an ordered event stream into current arc state. Pure, zero deps."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from .types import ArcStatusEvent, ArcStatusPhase


@dataclass(frozen=True, slots=True)
class ArcMilestone:
    title: str
    sent_at: str
    step: int | None = None
    total: int | None = None


@dataclass(frozen=True, slots=True)
class ArcBlocker:
    title: str
    sent_at: str
    body: str | None = None


@dataclass(frozen=True, slots=True)
class ArcState:
    arc_id: str
    #: Objective, from the ``started`` event (falls back to the first event).
    title: str
    #: Phase of the most recent event.
    phase: ArcStatusPhase
    #: Derived lifecycle status; ``blocked`` clears once a later non-blocked event arrives.
    status: Literal["active", "blocked", "done"]
    started_at: str
    last_event_at: str
    event_count: int
    milestones: list[ArcMilestone]
    #: The current blocker if the most recent event is ``blocked``, else ``None``.
    blocked: ArcBlocker | None
    step: int | None = None
    total: int | None = None
    eta_minutes: float | None = None


def _is_int(value: object) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def _is_number(value: object) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def reduce_arc(events: list[ArcStatusEvent]) -> ArcState | None:
    """Reduce an ordered sequence for one arc into current state, or ``None`` for
    an empty sequence. Does not validate; pass a sequence you trust."""
    if not events:
        return None

    first = events[0]
    last = events[-1]
    started = next((e for e in events if e["phase"] == "started"), None)

    milestones: list[ArcMilestone] = []
    step: int | None = None
    total: int | None = None
    eta_minutes: float | None = None

    for e in events:
        if e["phase"] == "milestone":
            milestones.append(
                ArcMilestone(
                    title=e["title"], sent_at=e["sent_at"], step=e.get("step"), total=e.get("total")
                )
            )
        s = e.get("step")
        if _is_int(s):
            step = s
        t = e.get("total")
        if _is_int(t):
            total = t
        et = e.get("eta_minutes")
        if _is_number(et):
            eta_minutes = et

    if last["phase"] == "done":
        status: Literal["active", "blocked", "done"] = "done"
    elif last["phase"] == "blocked":
        status = "blocked"
    else:
        status = "active"

    blocked = (
        ArcBlocker(title=last["title"], sent_at=last["sent_at"], body=last.get("body"))
        if last["phase"] == "blocked"
        else None
    )

    return ArcState(
        arc_id=first["arc_id"],
        title=started["title"] if started is not None else first["title"],
        phase=last["phase"],
        status=status,
        started_at=started["sent_at"] if started is not None else first["sent_at"],
        last_event_at=last["sent_at"],
        event_count=len(events),
        milestones=milestones,
        blocked=blocked,
        step=step,
        total=total,
        eta_minutes=eta_minutes,
    )
